### ISSUE TITLE
Compute policy gradient using cross entropy loss

### ISSUE THREAD

**Issue by weiliu620** [yukezhu/tensorflow-reinforce#10]
This is not an issue of the code per se, but I am learning RL and am wondering how the policy gradient is calculated in `pg_reinforce.py`. In this line 
```
self.cross_entropy_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(
logits=self.logprobs, labels=self.taken_actions)
```
A loss defined as the cross entropy between the action distribution from policy, and the actual action taken. But standard text such as Sutton define  the policy gradient in the the REINFORCE algorithm simply as the 'log grad' of the policy function. 

What is the difference of the two definitions? It seems cross entropy loss is a more general definition because it includes not just the action taken but all remaining actions. Can you give a reference of your method?

This is great project. Thank you for sharing this.

**Comment by Utsavz** [comment]
Did you get an answer? @weiliu620 Currently I am learning RL. Please help

**Comment by PaulKim1995** [comment]
I was also confused by the use of this cross-entropy loss, but in pg_actor_critic.py line 109. 

I don't know if this sort of loss function will work. The objective as written is to make the distribution of the logprobs of the actions calculated from by the policy to be identical to the distribution of the "taken_actions" which are fed from an action buffer that is populated by actions decided upon by the policy itself. Not sure what that does when executed. 

My guess is that he just reused old code from the cross-entropy method implementation elsewhere in the repo, without updating the loss function for different cases. In which case we should disregard this section, which is a shame because I really like the rest of his architecture and design. 

I am just starting to learn RL so please take what I say with a grain of salt

**Comment by luofuli** [comment]
@Utsavz @weiliu620 
- The 'log grad' of the policy function is the log(p(x)).
- The loss function of cross-entropy is \sum q * log(p).

These two are the same because q is a one-hot vector (the index of x is 1, others are 0). That's is two say,when the q is a one-hot vector, \sum q * log(p) = log(p(x)), where x refers to the sampled example.
